taichi/ir/ir_builder.cpp

```diff
@@ -10,7 +11,87 @@ IRBuilder::IRBuilder() {
 }
 
 Stmt *IRBuilder::insert(std::unique_ptr<Stmt> &&stmt) {
-  return insert_point_.block->insert(std::move(stmt), insert_point_.position++);
+  return insert(std::move(stmt), &insert_point_);
+}
+
+Stmt *IRBuilder::insert(std::unique_ptr<Stmt> &&stmt,
+                        InsertPoint *insert_point) {
+  return insert_point->block->insert(std::move(stmt), insert_point->position++);
+}
+
+void IRBuilder::set_insertion_point(InsertPoint new_insert_point) {
+  insert_point_ = new_insert_point;
+}
+
+void IRBuilder::set_insertion_point_to_after(Stmt *stmt) {
+  set_insertion_point({stmt->parent, stmt->parent->locate(stmt) + 1});
+}
+
+void IRBuilder::set_insertion_point_to_before(Stmt *stmt) {
+  set_insertion_point({stmt->parent, stmt->parent->locate(stmt)});
+}
+
+void IRBuilder::set_insertion_point_to_loop_begin(Stmt *loop) {
+  if (auto range_for = loop->cast<RangeForStmt>()) {
+    set_insertion_point({range_for->body.get(), 0});
+  } else if (auto struct_for = loop->cast<StructForStmt>()) {
+    set_insertion_point({struct_for->body.get(), 0});
+  } else if (auto while_stmt = loop->cast<WhileStmt>()) {
+    set_insertion_point({while_stmt->body.get(), 0});
+  } else {
+    TI_ERROR("Statement {} is not a loop.", loop->name());
+  }
+}
+
+void IRBuilder::set_insertion_point_to_true_branch(Stmt *if_stmt) {
+  TI_ASSERT(if_stmt->is<IfStmt>());
```

Comment:
Then we need to include each `XxxStmt`'s name in `ir.h`. Since we actually do that now, maybe it's fine with that?

And for convenience, maybe we want to set the return values of `IRBuilder::create_xxx` to the corresponding `XxxStmt` so that we can write `builder.set_insertion_point_to_true_branch(builder.create_if(...))` without more type casts.

Reply:
>  Since we actually do that now, maybe it's fine with that?

Hmm, could you share how long it takes for you to re-compile when `ir.h` is changed? (Also, how many threads are you using typically?) If it is too slow for you, would you take a look at `ir.h` and try to split it into smaller pieces? 

That said, for IRBuilder, by definition it will need to know all the stmts, so I'm not sure if/how we can avoid including the contents in `ir.h`.

> And for convenience, maybe we want to set the return values of IRBuilder::create_xxx to the corresponding XxxStmt so that we can write builder.set_insertion_point_to_true_branch(builder.create_if(...)) without more type casts.

Yeah, that sounds like a good idea :)

Reply:
> Hmm, could you share how long it takes for you to re-compile when ir.h is changed? (Also, how many threads are you using typically?)

It took me 4m17s just now with 8 threads. I think it's acceptable...